File: week_1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number = get_long("Number: ");
    int digit = 0;
    long digitNum = number;

    // check digit number
    while (digitNum / 10 > 0)
    {
        digitNum = digitNum / 10;
        digit++;
    }
    digit += 1;

    // sum 1
    long firstSumNum = number / 10;
    long firstSum = 0;
    long totalSum = 0;
    long SumDig = digit / 2;
    for (int i = 1; i <= SumDig; i++)
    {
        firstSum = firstSumNum % 10 * 2;
        firstSumNum = firstSumNum / 100;
        totalSum += firstSum / 10 + firstSum % 10;
    }

    // sum2
    long secondSum = 0;
    long secondSumNum = number;
    if (digit % 2 != 0)
    {
        SumDig++;
    }
    for (int i = 1; i <= SumDig; i++)
    {
        secondSum = secondSumNum % 10;
        secondSumNum = secondSumNum / 100;
        totalSum += secondSum;
    }

    // alt approach1 loop
    // int currentNumber = number;
    // int sum1 = 0;
    // int sum2 = 0;
    // for (int i = 0; i < digit; i++)
    // {
    //     int currentDigit = currentNumber % 10;
    //     if(i % 2 == 0)
    //     {
    //         // even digit
    //         sum1 += currentDigit;
    //     }
    //     else
    //     {
    //         // odd digit
    //         int multi = currentDigit * 2;
    //         sum2 += multi / 10;
    //         sum2 += multi % 10;
    //     }
    // }
    // int finalSum = sum1 + sum2;

    // cheking card type
    if (totalSum % 10 == 0)
    {
        long creditNum15_2 = number / 10000000000000;
        long creditNum16_2 = number / 100000000000000;
        if (digit == 15)
        {
            if (creditNum15_2 == 34 || creditNum15_2 == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (digit == 16)
        {
            if (number / 1000000000000000 == 4)
            {
                printf("VISA\n");
            }
            else if (creditNum16_2 == 51 || creditNum16_2 == 52 || creditNum16_2 == 53 || creditNum16_2 == 54 ||
                     creditNum16_2 == 55)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else if (digit == 13)
        {
            if (number / 1000000000000 == 4)
            {
                printf("VISA\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: week_6/sentimental-credit/credit.py
from cs50 import get_string


def main():
    while True:
        number = get_string("Number: ")
        if (number.isdecimal()):
            if (len(number) == 13) or (len(number) == 15) or (len(number) == 16):
                sums = luhn_Algorithm(number)
                if sums % 10 == 0:
                    check_comp(number)
                else:
                    print("INVALID")
            else:
                print("INVALID")
            break


def luhn_Algorithm(number):
    sums = 0
    number = number[::-1]
    for n in range(1, len(number), 2):
        num1, num2 = (int(number[n]) * 2) // 10, (int(number[n]) * 2) % 10
        # print("number[n]: ",number[n])
        # print(f'num1: {num1}, num2: {num2}')
        sums += num1 + num2
    for n in range(0, len(number), 2):
        sums += int(number[n])

    return sums


def check_comp(number):
    if (len(number) == 15) and (int(number[0]) == 3):
        if (int(number[1]) == 4) or (int(number[1]) == 7):
            print("AMEX")
        else:
            print("INVALID")

    elif (len(number) == 16):
        if (int(number[0]) == 5):
            if (int(number[1]) == 1) or (int(number[1]) == 2) or (int(number[1]) == 3) or (int(number[1]) == 4) or (int(number[1]) == 5):
                print("MASTERCARD")
            else:
                print("INVALID")

        elif (int(number[0]) == 4):
            print("VISA")

        else:
            print("INVALID")

    elif (len(number) == 13) and (int(number[0]) == 4):
        print("VISA")

    else:
        print("INVALID")


main()
